### DISCUSSION TITLE
Search Filters when Fetching Entities using the Aggregate to Avoid Manual Filtering

### DISCUSSION THREAD

**Discussion by abhirathmahipal** [proteanhq/protean#370] [Ideas]
Search filters when fetching entities using the aggregate to avoid having to write loops (manually filter) or directly using `dao` of the entity.

Relevant Links
- [Backpopulates](https://medium.com/@ns2586/sqlalchemys-relationship-and-lazy-parameter-4a553257d9ef)
- [An Example with Filter](https://stackoverflow.com/a/8562155)

Let's take a simple example of `Entry` and `Members`. To get the list of active members, I'm forced to write a `repo` method.

### Simple Example
```python
@domain.aggregate
class Entry:
    members = HasMany(Members)


# partial repo code
@classmethod
 def fetch_members_of_entry(cls, entry_id):
     return current_domain.get_dao(Member).find_by(entry_id=entry_id, is_archived=True)

```

I could do something like `entry.members` to get all the members. This is not expressive enough in my opinion.

I suggest something in lines of
We already have methods like `add_`, `remove_` which get created dynamically for every reference. Why not create another method called `get_{reference_name}`?

### Proposed Solution

```python
entry = entry_repo.get("a27bf445-1403-468e-a5bf-f1f2a9631b8b")
entry.get_members(is_archived=True)

```

### A More Involved Example Showcasing the Problem

```python
 @classmethod
    def fetch_pca_finalists(cls, competition_id):
        award_dao = current_domain.get_dao(Award)
        award = award_dao.find_by(
            competition_id=competition_id, default_peoples_choice_award=True
        )
        finalist_pca_award = next(
            (
                award_type
                for award_type in award.award_types
                if award_type.award_type == AwardPosition.FINALIST.value
            ),
            None,
        )
        result = []
        if finalist_pca_award:
            for awarded_entry in award.awarded_entries:
                if awarded_entry.award_type_id == finalist_pca_award.id:
                    result.append(awarded_entry.entry)

        return award, result

```

**Comment by subhashb** [comment]
Following up on this a while later. Good news: this is already in Protean, and it landed pretty much exactly as you proposed.

Every `HasMany` field now gets two extra helper methods alongside `add_` and `remove_`:

- `filter_<field>(**kwargs)`: returns all linked entities matching the criteria
- `get_one_from_<field>(**kwargs)`: returns a single matching entity, and raises `ObjectNotFoundError` / `TooManyObjectsError` if there's zero or more than one match

For your `Entry`/`Members` example, that means:

```python
@domain.aggregate
class Entry:
    members = HasMany(Member)

entry = entry_repo.get("a27bf445-1403-468e-a5bf-f1f2a9631b8b")
active_members = entry.filter_members(is_archived=True)
```

No repo method, no DAO, no manual loop.

For your award/finalist example, this cleans it up nicely too:

```python
@classmethod
def fetch_pca_finalists(cls, competition_id):
    award_dao = current_domain.get_dao(Award)
    award = award_dao.find_by(
        competition_id=competition_id, default_peoples_choice_award=True
    )

    finalist_pca_award = award.get_one_from_award_types(
        award_type=AwardPosition.FINALIST.value
    )

    result = [
        awarded_entry.entry
        for awarded_entry in award.filter_awarded_entries(
            award_type_id=finalist_pca_award.id
        )
    ]
    return award, result
```

One thing worth calling out so it doesn't surprise you: `filter_` and `get_one_from_` operate on the collection already loaded into memory on the aggregate, not a fresh database query. Protean loads the full `HasMany` collection when you fetch the aggregate, and these methods just filter over that in-Python. For most aggregate-sized collections that's exactly what you want (no extra round trip), but if you're dealing with a child collection large enough that you don't want it fully loaded at all, that's a case for a dedicated DAO query instead.

Docs are here if you want the full rundown: [Auto-Generated Helper Methods](https://docs.proteanhq.com/guides/domain-definition/relationships/#auto-generated-helper-methods).

Let us know if this covers what you had in mind, or if there's a case it doesn't handle. Happy to dig further.
